src/main/java/org/jenkinsci/plugins/gitclient/GitUsernamePasswordBinding.java

```diff
@@ -0,0 +1,158 @@
+    final static private String GIT_PASSWORD_KEY = "GIT_PASSWORD";
+    private final Map<String, String> credMap = new LinkedHashMap<>();
+    private String gitTool = null;
+
+    @DataBoundConstructor
+    public GitUsernamePasswordBinding(String credentialsId) {
+        super(credentialsId);
+        //Variables could be added if needed
+    }
+
+    @Override
+    protected Class<StandardUsernamePasswordCredentials> type() {
+        return StandardUsernamePasswordCredentials.class;
+    }
+
+    @Override
+    public MultiEnvironment bind(@NonNull Run<?, ?> run, FilePath filePath,
+                                 Launcher launcher, @NonNull TaskListener taskListener)
+            throws IOException, InterruptedException {
+        StandardUsernamePasswordCredentials credentials = getCredentials(run);
+        setKeyBindings(credentials);
+        gitTool = gitToolName(run, taskListener);
+        if (gitTool != null && filePath != null) {
+            final UnbindableDir unbindTempDir = UnbindableDir.create(filePath);
+            setRunEnvironmentVariables(filePath, taskListener);
+            GenerateGitScript gitScript = new GenerateGitScript(credentials.getUsername(),
+                    credentials.getPassword().getPlainText(), credentials.getId());
+            FilePath gitTempFile = gitScript.write(credentials, unbindTempDir.getDirPath());
+            credMap.put("GIT_ASKPASS", gitTempFile.getRemote());
+            return new MultiEnvironment(credMap, unbindTempDir.getUnbinder());
+        } else {
+            return new MultiEnvironment(credMap);
+        }
+    }
+
+    @Override
+    public Set<String> variables() {
+        Set<String> keys = new LinkedHashSet<>();
+        keys.add(GIT_USERNAME_KEY);
+        keys.add(GIT_PASSWORD_KEY);
+        return keys;
+    }
+
+
+    @Override
+    public void setKeyBindings(@NonNull StandardCredentials credentials) {
+        credMap.put(GIT_USERNAME_KEY, ((StandardUsernamePasswordCredentials) credentials).getUsername());
+        credMap.put(GIT_PASSWORD_KEY, ((StandardUsernamePasswordCredentials) credentials).getPassword().getPlainText());
+    }
+
+    @Override
+    public void setRunEnvironmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException {
+        if (!Functions.isWindows() && ((CliGitAPIImpl) getGitClientInstance(listener)).
+                isAtLeastVersion(2, 3, 0, 0)) {
+            credMap.put("GIT_TERMINAL_PROMPT", "false");
+        } else {
+            credMap.put("GCM_INTERACTIVE", "false");
+        }
+    }
+
+    @Override
+    public GitClient getGitClientInstance(TaskListener listener) throws IOException, InterruptedException {
+        Git gitInstance = Git.with(listener, new EnvVars()).using(gitTool);
+        return gitInstance.getClient();
+    }
+
+    protected static final class GenerateGitScript extends AbstractOnDiskBinding<StandardUsernamePasswordCredentials> {
```

Comment:
am I missing somethign here that informs the credential plugin that  the actual username and password should be masked?

If I ran GIT_ASK_PASS from the CLI (outside of git) would the password be revealed in plain text?

Reply:
As far as I can tell, the credentials binding plugin classes provide the masking of the username and password.  At least that is what I've seen in my interactive testing.